### PR TITLE
Activate NodeIPAM for OpenShift platform

### PR DESCRIPTION
**Pull request by ksamoray**
OpenShift does not run K8S NodeIPAM controller by default, so the operator should activate NodeIPAM within the Antrea Controller and configure it with the CIDR data from the cluster configuration.

### CODE REVIEW

File: controllers/config/config.go [hunk 1]
```diff
@@ -56,10 +60,6 @@ func fillConfig(clusterConfig *configv1.Network, operConfig *operatorv1.AntreaIn
 	if !ok {
 		antreaAgentConfig[types.DefaultMTUOption] = types.DefaultMTU
 	}
-	// Set Antrea image.
-	if operConfig.Spec.AntreaImage == "" {
-		operConfig.Spec.AntreaImage = types.DefaultAntreaImage
-	}
```
Comment (salv-orlando) [inline controllers/config/config.go:62]: Is this change unrelated to this PR?

File: controllers/config/config.go [hunk 2]
```diff
@@ -68,12 +68,94 @@ func fillConfig(clusterConfig *configv1.Network, operConfig *operatorv1.AntreaIn
...
+
+	if clusterConfig != nil {
+		ip4found := false
+		ip6found := false
+
+		for _, cidr := range clusterConfig.Spec.ClusterNetwork {
+			if net.IsIPv4CIDRString(cidr.CIDR) {
+				if !ip4found {
+					ip4found = true
+					controllerConfig.NodeIPAM.ClusterCIDRs = append(controllerConfig.NodeIPAM.ClusterCIDRs, cidr.CIDR)
```
Comment (salv-orlando) [inline controllers/config/config.go:93]: I am not sure I understand the logic around the `ip4found` and `ip6found` variables. Can you explain?

File: controllers/config/config.go [hunk 3]
```diff
@@ -68,12 +68,94 @@ func fillConfig(clusterConfig *configv1.Network, operConfig *operatorv1.AntreaIn
...
+			} else {
+				if !ip6found {
+					ip6found = true
+					controllerConfig.NodeIPAM.ServiceCIDRv6 = svcCIDR
+				}
+			}
+		}
+	}
+
+	updatedAntreaControllerConfig, err := yaml.Marshal(controllerConfig)
```
Comment (salv-orlando) [inline controllers/config/config.go:124]: Probably still me being obtuse, but I'm not sure why we need to copy structs from Antrea from marshalling and unmarhsalling. I think the existing fillConfig code makes already changes that ultmately end up in Antrea agent config.

File: deploy/openshift/operator.antrea.vmware.com_v1_antreainstall_cr.yaml [hunk 4]
```diff
@@ -193,5 +193,5 @@ spec:
 
     # TLS min version from: VersionTLS10, VersionTLS11, VersionTLS12, VersionTLS13.
     #tlsMinVersion:
-  antreaImage: projects.registry.vmware.com/antrea/antrea-ubi:v0.13.1
+  antreaImage: projects.registry.vmware.com/antrea/antrea-ubi:latest
```
Comment (salv-orlando) [inline deploy/openshift/operator.antrea.vmware.com_v1_antreainstall_cr.yaml:196]: Isnt the same change also in PR #45 ?

File: controllers/config/config.go (hunk 1 above)
Comment (ksamoray) [reply controllers/config/config.go:62]: Moved to line 146 within fillConfig() as it's not agent nor controller param

File: controllers/config/config.go (hunk 2 above)
Comment (ksamoray) [reply controllers/config/config.go:93]: clusterConfig.Spec.ClusterNetwork contains a list of IP4, IP6 CIDRs.
I would like to take the first IP4 and first IP6 CIDRs from the list and append these into the controller config.
I could basically assign the while list from the cluster to the controller config but I need their matching prefix lengths as well.

File: controllers/config/config.go (hunk 3 above)
Comment (ksamoray) [reply controllers/config/config.go:124]: This is basically how it's done in #47 - not that the controller config and NodeIPAM within is hierarchical and therefore is more cumbersome to implement with maps of interfaces.
And the more params we want to set within the configs, the worse it becomes (and requires more effort to revert).

File: deploy/openshift/operator.antrea.vmware.com_v1_antreainstall_cr.yaml (hunk 4 above)
Comment (ksamoray) [reply deploy/openshift/operator.antrea.vmware.com_v1_antreainstall_cr.yaml:196]: Indeed, it's redundant here

File: controllers/config/config.go (hunk 2 above)
Comment (salv-orlando) [reply controllers/config/config.go:93]: Thanks for the clarification. Will this code eventually use all the CIDRs in `clusterConfig.Spec.ClusterNetwork`... I think we would like to make sure all of them are used

File: controllers/config/config.go (hunk 3 above)
Comment (salv-orlando) [reply controllers/config/config.go:124]: I figured out the additional challenge after looking at the code in PR #47. 
I see the implementation here is more straightforward. The downside is about having the additional hassle of periodically syncrhonizing structs for Antrea Config.

File: controllers/config/config.go (hunk 2 above)
Comment (ksamoray) [reply controllers/config/config.go:93]: Not sure whether this is a limitation applied by Antrea Agent or by NodeIPAM itself. So it doesn't matter what this code does but how Antrea or NodeIPAM utilizes these CIDRs.

File: controllers/config/config.go (hunk 3 above)
Comment (ksamoray) [reply controllers/config/config.go:124]: Maybe if we choose to take this approach, it's worthwhile to add a periodic CI job which validates the config structs against Antrea.

**Review by salv-orlando — APPROVED** [review]
Let's adopt this approach

**Pull request merged**